### ISSUE TITLE
 cas_io_cache1_1 Tainted in kernel case kernel panic

### ISSUE THREAD

**Issue by perrynzhou** [Open-CAS/open-cas-linux#342]
@mmichal10 @robertbaldyga  we find opencas service case kernel panic in our product env,detail info as fellows,is exists some bugs in opencas?

- os env
```
3.10.0-693.el7.x86_64
```
- opencas version

```
[root@10-193-69-43.v-bj-4.vivo.lan:/root]
# casadm --version
+=========================+===================+
| Name                    |      Version      |
+=========================+===================+
| CAS Cache Kernel Module | 19.09.00.00001210 |
| CAS Disk Kernel Module  | 19.09.00.00001210 |
| CAS CLI Utility         | 19.09.00.00001210 |
+=========================+===================+

[root@10-193-69-43.v-bj-4.vivo.lan:/root]
# casadm -L
type    id   disk       status    write policy   device
cache   1    /dev/sdc   Running   wt             -
+core   1    /dev/sdb   Active    -              /dev/cas1-1

```
- kernel logs
```
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405271] ------------[ cut here ]------------
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405297] WARNING: CPU: 10 PID: 39797 at lib/list_debug.c:53 __list_del_entry+0x63/0xd0
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405299] list_del corruption, ffff885fac756ff0->next is LIST_POISON1 (dead000000000100)
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405300] Modules linked in: fuse binfmt_misc cas_cache(OE) cas_disk(OE) tcp_diag inet_diag dm_mirror dm_region_hash dm_log dm_mod iTCO_wdt iTCO_vendor_support sb_edac edac_core intel_powerclamp coretemp intel_rapl iosf_mbi kvm_intel kvm irqbypass crc32_pclmul ghash_clmulni_intel aesni_intel lrw gf128mul glue_helper ablk_helper cryptd sg pcspkr hpilo hpwdt ioatdma shpchp i2c_i801 lpc_ich dca ipmi_si ipmi_devintf ipmi_msghandler pcc_cpufreq wmi acpi_power_meter ip_tables xfs sd_mod crc_t10dif crct10dif_generic mgag200 i2c_algo_bit drm_kms_helper bnx2x syscopyarea sysfillrect sysimgblt fb_sys_fops ttm drm crct10dif_pclmul crct10dif_common crc32c_intel tg3 serio_raw hpsa mdio i2c_core ptp pps_core scsi_transport_sas libcrc32c
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405335] CPU: 10 PID: 39797 Comm: cas_io_cache1_1 Tainted: G        W  OE  ------------   3.10.0-693.el7.x86_64 #1
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405337] Hardware name: HP ProLiant DL380 Gen9/ProLiant DL380 Gen9, BIOS P89 02/17/2017
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405338]  ffff882fbda7bc88 000000004fed9b41 ffff882fbda7bc38 ffffffff816a3d91
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405340]  ffff882fbda7bc78 ffffffff810879c8 00000035bda7bce0 ffff885fac756ff0
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405341]  ffff885fac756000 ffff885fb9c8b2e0 0000000000000282 ffff885fbbf12ab8
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405344] Call Trace:
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405352]  [<ffffffff816a3d91>] dump_stack+0x19/0x1b
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405357]  [<ffffffff810879c8>] __warn+0xd8/0x100
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405360]  [<ffffffff81087a4f>] warn_slowpath_fmt+0x5f/0x80
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405370]  [<ffffffffc05ba3de>] ? cas_rpool_try_get+0x5e/0x80 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405371]  [<ffffffff8133d963>] __list_del_entry+0x63/0xd0
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405373]  [<ffffffff8133d9dd>] list_del+0xd/0x30
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405377]  [<ffffffffc05ba3cf>] cas_rpool_try_get+0x4f/0x80 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405382]  [<ffffffffc05bf08d>] __cas_ctx_data_alloc+0xed/0x300 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405387]  [<ffffffffc05bf2b0>] cas_ctx_data_alloc+0x10/0x20 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405393]  [<ffffffffc05c8163>] _ocf_read_generic_do+0x1a3/0x2b0 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405397]  [<ffffffffc05c8348>] ocf_read_generic+0xd8/0x100 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405405]  [<ffffffffc05e0813>] ocf_io_handle+0x23/0x40 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405413]  [<ffffffffc05e0875>] ocf_queue_run_single+0x45/0x50 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405419]  [<ffffffffc05e08a8>] ocf_queue_run+0x28/0x50 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405424]  [<ffffffffc05b7d9a>] _cas_io_queue_thread+0xfa/0x150 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405429]  [<ffffffff810b1910>] ? wake_up_atomic_t+0x30/0x30
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405434]  [<ffffffffc05b7ca0>] ? cas_service_ioctl_ctrl+0x1bd0/0x1bd0 [cas_cache]
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405436]  [<ffffffff810b098f>] kthread+0xcf/0xe0
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405438]  [<ffffffff810b08c0>] ? insert_kthread_work+0x40/0x40
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405442]  [<ffffffff816b4f18>] ret_from_fork+0x58/0x90
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405444]  [<ffffffff810b08c0>] ? insert_kthread_work+0x40/0x40
Feb 28 09:46:14 10-193-69-43 kernel: [6872898.405446] ---[ end trace 808e8663e223caae ]---
Feb 28 09:46:14 10-193-69-43 k
```

**Comment by mmichal10** [comment]
Hi @perrynzhou, thank you for reporting this issue.

Could you provide us more information? It occurred only once? You were doing any particular management operation ( starting cache, changing cache mode etc) or CAS was just running in background? Would there be a possibility to retest it on CAS v20.1?

**Comment by perrynzhou** [comment]
> Hi @perrynzhou, thank you for reporting this issue.
> 
> Could you provide us more information? It occurred only once? You were doing any particular management operation ( starting cache, changing cache mode etc) or CAS was just running in background? Would there be a possibility to retest it on CAS v20.1?

yes,just occurred once within running long time.at same time,cas was running in background when started.we used cas 19.3.we find some problem about opencas:

- when we execute casctl start ,why cost much cpu resource?that case no response  for node?

**Comment by imjfckm** [comment]
Hi @perrynzhou,
### First issue (kernel panic):
I see from logs that you're using CAS 19.9, but you say you used 19.3, why is that?
We would need some more details:
* what is the workload that's being run on the cas device?
* what type and size of devices you're using
* what are the machine specs (RAM, CPU)
* what is the context of this panic? did kernel complain about lack of memory? (oom killer logs)

### Second issue (casctl start being CPU intensive):
Please file **separate issue** for that. But in this case we also need some more details (please include them when you file this new issue):
* what is the CAS configuration (disk types and sizes, cache/core configuration)
* is the node unresponsive only for some time or is it dead for good? if it's just temporary - how long does it take to complete?
* what are the machine specs (RAM, CPU)

**Comment by perrynzhou** [comment]
@imjfckm thanks you 
- when we found that ,the node is crash,that can't  collected some runtime resource
- 4k block ssd device and 7T  hdd for that

**Comment by imjfckm** [comment]
Hi @perrynzhou,

I see you still haven't submitted an issue regarding `casctl start` - I'd really appreciate if you did that, we could then separate the talks about both problems you're encountering.

Aside from that I don't really understand what do you mean when you say "can't collected some runtime resource", and to which issue it pertains.

Also, what do you mean by 4k block ssd device? Is it a SSD drive that's 4 KiB in size? Or is it sector size? How big is the cache device?

You didn't include the configuration you're using, and workload you're using, so right now we don't know even where to start with reproducing and fixing your issues.

**Comment by perrynzhou** [comment]
@imjfckm  this is syste info ,include kernel version,opencas version ,disk info,opencas config.
```
[root@ceph-10-193-69-43.v-bj-4.vivo.lan:/etc/opencas]
# uname -a
Linux ceph-10-193-69-43.v-bj-4.vivo.lan 3.10.0-693.el7.x86_64 #1 SMP Tue Aug 22 21:09:27 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
# fdisk -l|grep sd
Disk /dev/sda: 600.1 GB, 600092704768 bytes, 1172056064 sectors
/dev/sda1   *        2048     1050623      524288   83  Linux
/dev/sda2         1050624  1138501631   568725504   83  Linux
/dev/sda3      1138501632  1172056063    16777216   82  Linux swap / Solaris
Disk /dev/sdb: 40007.5 GB, 40007525990400 bytes, 78139699200 sectors
Disk /dev/sdc: 1920.3 GB, 1920349503488 bytes, 3750682624 sectors

[root@ceph-10-193-69-43.v-bj-4.vivo.lan:/etc/opencas]
#  lsblk -d -o name,rota
NAME ROTA
sda     1
sdb     1
sdc     0
[root@ceph-10-193-69-43.v-bj-4.vivo.lan:/etc/opencas]
# cat opencas.conf 
version=19.3.0
## Caches configuration section
[caches]
/dev/sdc       WT
1               /dev/disk/by-id/wwn-0x600508b1001c54069e065e7aa64af034  WT

[cores]
1              1               /dev/disk/by-id/wwn-0x123456789abcdef0
/dev/sdb
1               1               /dev/disk/by-id/wwn-0x600508b1001c97b93e60e10ac233017a
```

**Comment by imjfckm** [comment]
Hi @perrynzhou, thanks for providing the config. 
* Judging from the hostname I would assume that this machine is an OSD node for Ceph cluster, would that be correct?
* If this is indeed Ceph - what is the configuration? What is the Ceph version?
* What kind of workload was run on the node/cluster? 

We need more info to even try and reproduce the failure you're seeing.

**Comment by perrynzhou** [comment]
@imjfckm  no ceph installtion, that node just named ceph

**Comment by imjfckm** [comment]
We do have some longevity testing on premises and didn't find issues similar to yours. Without any other details on how we might reproduce this particular issue I don't see how can we approach finding and fixing it.

In the meantime we introduced some runtime checks to make debugging issues like that easier in the future https://github.com/Open-CAS/open-cas-linux/commit/d822a1d12e7b7131f8123baaddd6f5b9595bcd3f. It's already merged in, but I suggest waiting for the next official release to update. If the problem is indeed there, it should be easier to spot now and find the root cause.

**Comment by eriksun421** [comment]
Hi @perrynzhou , are you able to reproduce this issue with latest v20.3 release, which has a lot of bug fixes ? Can you please provide following information if you are still able to reproduce it on v20.3 release ?

1. OS/kernel version (uname -a)
2. casadm -V
3. casadm -L
4. casadm -P -i X
5. cat /etc/opencas/opencas.conf
6. lsblk
6. complete demsg log
7. Detailed steps to reproduce this issue. If running FIO test, please also share FIO test command and config file.

**Comment by mmichal10** [comment]
I'm closing issue due the lack of response
